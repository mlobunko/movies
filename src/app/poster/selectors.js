import { NAME } from './constants';

export const getPoster = state => state[NAME];
