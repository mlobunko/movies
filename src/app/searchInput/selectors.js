import { NAME } from './constants';

export const getBackdropPathFromData = data => data['backdrop_path'];

export const getErrorFromState = state => state[NAME]['error'];

export const getIdFromSelected = result => '' + result['id'];

export const getList = arr => {
  let listString = '';
  arr.forEach(
    (el, i) =>
      i === 0 ? (listString += `${el.name}`) : (listString += `, ${el.name}`)
  );
  return listString;
};

export const getDescriptionFromData = data => {
  const description = {};

  description.title = data['title'];
  description.tagline = data['tagline'];
  description.overview = data['overview'];
  description.genres = getList(data['genres']);
  description.productionCompanies = getList(data['production_companies']);
  description.releaseDate = data['release_date'];
  description.runtime = '' + data['runtime'];
  description.revenue = data['revenue'].toLocaleString('en');
  description.voteAvarage = '' + data['vote_average'];

  return description;
};

export const getPosterPathFromData = data => data['poster_path'];

export const getSearchResultFromData = data => data['results'].slice(0, 5);

export const getSearchResultFromState = state => state[NAME]['searchResult'];
