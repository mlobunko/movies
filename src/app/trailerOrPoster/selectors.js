import { NAME } from './constants';

export const getPoster = state => state[NAME]['posterPath'];

export const getTrailer = state => state[NAME]['trailerPath'];
