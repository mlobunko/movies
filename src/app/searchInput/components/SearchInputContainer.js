import React from 'react';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import { connect } from 'react-redux';

import RenderSuggestion from './RenderSuggestion';
import * as actions from '../actions';
import * as selectors from '../selectors';
import { StyledContainer, theme } from './styles';

export class SearchInputContainer extends React.Component {
  state = {
    value: '',
    suggestions: []
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.suggestions !== prevState.suggestions) {
      return {
        suggestions: nextProps.suggestions
      };
    } else return null;
  }

  onChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  onSuggestionsFetchRequested = ({ value }) => {
    const valueForSearch = value.trim();
    if (valueForSearch.length >= 2) {
      this.props.makeRequestForSearch(valueForSearch);
    }
  };

  onSuggestionSelected = (event, { suggestion }) => {
    this.props.makeRequestForDetails(selectors.getIdFromSelected(suggestion));
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };
  render() {
    const { value, suggestions } = this.state;
    const inputProps = {
      placeholder: 'Search Movie Title...',
      value,
      onChange: this.onChange
    };
    return (
      <StyledContainer>
        <Autosuggest
          suggestions={suggestions}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          onSuggestionSelected={this.onSuggestionSelected}
          getSuggestionValue={selectors.getSuggestionValue}
          renderSuggestion={RenderSuggestion}
          inputProps={inputProps}
          theme={theme}
        />
      </StyledContainer>
    );
  }
}

const mapStateToProps = state => ({
  getSuggestionValue: selectors.getSuggestionValue,
  suggestions: selectors.getSuggestionsFromState(state)
});

const mapDispatchToProps = {
  makeRequestForDetails: actions.makeRequestForDetails,
  makeRequestForSearch: actions.makeRequestForSearch,
  setSuggestions: actions.setSuggestions
};

SearchInputContainer.displayName = 'SearchInputContainer';

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchInputContainer);

SearchInputContainer.propTypes = {
  getSuggestionValue: PropTypes.func.isRequired,
  suggestions: PropTypes.array.isRequired,
  makeRequestForDetails: PropTypes.func.isRequired,
  makeRequestForSearch: PropTypes.func.isRequired,
  setSuggestions: PropTypes.func.isRequired
};
