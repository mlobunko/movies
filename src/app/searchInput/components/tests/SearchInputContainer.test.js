import React from 'react';
import { shallow } from 'enzyme';

import { SearchInputContainer } from '../SearchInputContainer';
import * as selectors from '../../selectors';

jest.mock('../../selectors', () => ({
  getIdFromSelected: jest.fn(),
  getSuggestionValue: jest.fn()
}));

const initialProps = {
  getSuggestionValue: jest.fn(),
  suggestions: [],
  makeRequestForDetails: jest.fn(),
  makeRequestForSearch: jest.fn(),
  setSuggestions: jest.fn()
};

const initialState = {
  value: '',
  suggestions: []
};

beforeEach(() => {
  jest.resetAllMocks();
});

describe('render', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<SearchInputContainer {...initialProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('methods', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('getDerivedStateFromProps', () => {
    it('change state if new suggestions are not equal to old state', () => {
      const newState = {
        value: '',
        suggestions: ['a', 'b']
      };
      const newProps = {
        ...initialProps,
        suggestions: newState.suggestions
      };
      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.setProps(newProps);
      expect(wrapper.state()).toEqual(newState);
    });
  });

  describe('onChange', () => {
    it('change state', () => {
      const newState = {
        ...initialState,
        value: '123'
      };
      const e = {};
      const obj = {
        newValue: newState.value
      };
      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.instance().onChange(e, obj);
      expect(wrapper.state()).toEqual(newState);
    });
  });

  describe('onSuggestionsFetchRequested', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    it('dont call makeRequestForSearch, value.length less than 2', () => {
      const inputValue = {
        value: ' 1 '
      };
      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.instance().onSuggestionsFetchRequested(inputValue);
      expect(
        wrapper.instance().props.makeRequestForSearch
      ).toHaveBeenCalledTimes(0);
    });

    it('call makeRequestForSearch with right argument', () => {
      const inputValue = {
        value: ' 123 '
      };
      const expectedArg = '123';

      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.instance().onSuggestionsFetchRequested(inputValue);
      expect(
        wrapper.instance().props.makeRequestForSearch
      ).toHaveBeenCalledWith(expectedArg);
    });
  });

  describe('onSuggestionSelected', () => {
    it('call makeRequestForDetails with right argument', () => {
      selectors.getIdFromSelected.mockReturnValue('a');
      const e = {};
      const inputValue = {
        suggestion: 'a'
      };
      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.instance().onSuggestionSelected(e, inputValue);
      expect(
        wrapper.instance().props.makeRequestForDetails
      ).toHaveBeenCalledWith('a');
    });
  });

  describe('onSuggestionsClearRequested', () => {
    it('change suggestions in state to empty array', () => {
      const state = {
        ...initialState,
        suggestions: ['a', 'b', 'c']
      };
      const wrapper = shallow(<SearchInputContainer {...initialProps} />);
      wrapper.setState(state);
      wrapper.instance().onSuggestionsClearRequested();
      expect(wrapper.state().suggestions).toEqual([]);
    });
  });
});
