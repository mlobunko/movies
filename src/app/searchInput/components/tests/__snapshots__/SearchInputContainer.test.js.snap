// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<StyledContainer>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "aria-label": "search movies",
        "onChange": [Function],
        "placeholder": "Search Movie Title...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": Object {
          "position": "relative",
        },
        "input": Object {
          "background": "rgba(0, 0, 0, 0)",
          "border": "none",
          "borderBottom": "1px solid white",
          "color": "white",
          "fontSize": 16,
          "height": 30,
          "padding": "10px 20px",
          "width": "100%",
        },
        "inputFocused": Object {
          "outline": "none",
        },
        "inputOpen": Object {
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
        },
        "suggestion": Object {
          "cursor": "pointer",
          "padding": "10px 20px",
        },
        "suggestionHighlighted": Object {
          "backgroundColor": "#fff",
          "color": "#000",
        },
        "suggestionsContainer": Object {
          "display": "none",
        },
        "suggestionsContainerOpen": Object {
          "background": "rgba(0, 0, 0, 0.8)",
          "borderBottomLeftRadius": 4,
          "borderBottomRightRadius": 4,
          "display": "block",
          "position": "absolute",
          "width": "100%",
          "zIndex": 2,
        },
        "suggestionsList": Object {
          "listStyleType": "none",
          "margin": 0,
          "padding": 0,
        },
      }
    }
  />
</StyledContainer>
`;
